connect/runtime/src/test/java/org/apache/kafka/connect/integration/OffsetsApiIntegrationTest.java

```diff
@@ -895,6 +909,55 @@ private Map<String, String> baseSourceConnectorConfigs() {
         return props;
     }
 
+    /**
+     * Modify (i.e., alter or reset) the offsets for a sink connector, with retry logic to
+     * handle cases where laggy task shutdown may have left a consumer in the group.
+     * @param offsetsToAlter the offsets to alter for the sink connector, or null if
+     *                       the connector's offets should be reset instead
+     * @return the response from the REST API, if the request was successful
+     * @throws InterruptedException if the thread is interrupted while waiting for a
+     * request to modify the connector's offsets to succeed
+     * @see <a href="https://issues.apache.org/jira/browse/KAFKA-15826">KAFKA-15826</a>
+     */
+    private String modifySinkConnectorOffsetsWithRetry(ConnectorOffsets offsetsToAlter) throws InterruptedException {
+        // Some retry logic is necessary to account for KAFKA-15826,
+        // where laggy sink task startup/shutdown can leave consumers running
```

Comment:
I'm +1 for the retry logic here since I've seen the same occasional CI flakiness that you've indicated, but I don't quite get how it's related to `KAFKA-15826` given that the `MonitorableSinkConnector` / `MonitorableSinkTask` being used here isn't really expected to block long enough in its stop method to cause task cancellation to occur? Furthermore, given that the referenced ticket isn't resolved yet, retrying here wouldn't make sense since even task cancellation wouldn't cause the task's consumer to be closed right? Or is it linked here as more of a future reference (once the ticket is resolved)?

Reply:
The hypothesis I have is that our CI environment is so slow, that even `MonitorableSinkTask` instances may fail to shut down gracefully. And, because of KAFKA-15826, those task instances will still have running consumers that can interfere with requests to reset/alter consumer group offsets.

If my guess is correct, the retry logic should help because it gives the task a little more time to complete shut down and close its consumer, and it should no longer be necessary once KAFKA-15826 is resolved.

It's also possible that slow rebalances or metadata updates may cause transient errors even when all tasks have completed gracefully, though I have to admit I haven't done enough research on that part of the code base to speak with any certainty.

Does that clear things up?

Reply:
Okay fair enough, thanks for the explanation!